src/clients/pagination/stream.rs

```diff
@@ -9,7 +9,34 @@ use futures::{future::Future, stream::Stream};
 /// Alias for `futures::stream::Stream<Item = T>`, since async mode is enabled.
 pub type Paginator<'a, T> = Pin<Box<dyn Stream<Item = T> + 'a>>;
 
+pub type RequestFuture<'a, T> = Pin<Box<dyn 'a + Future<Output = ClientResult<Page<T>>>>>;
+
 /// This is used to handle paginated requests automatically.
+pub fn paginate_with_ctx<'a, Ctx: 'a, T, Request>(
+    ctx: Ctx,
+    req: Request,
+    page_size: u32,
+) -> Paginator<'a, ClientResult<T>>
+where
+    T: 'a + Unpin,
+    Request: 'a + for<'ctx> Fn(&'ctx Ctx, u32, u32) -> RequestFuture<'ctx, T>,
```

Comment:
I am not sure if I understand the design correctly, so I try to retell the story.

What the signature means is to allow `req` function to hold a reference to `Ctx` in the returned future with [HRTB](https://doc.rust-lang.org/nomicon/hrtb.html) 

So that the returned value of the `req` function is allowed to reference the argument(`Ctx`), we pass it.

All the thing we want to do is to specify that the lifetime of `Ctx` parameter must outlive the lifetime of `RequestFuture`?

Reply:
That sounds correct. I'll add my version, so you can double check. :)

The problem we need to solve is: When we move anything into the closure (`req`), the returned future must not reference it in some way, since we can't ensure that the closure outlives the returned future.

The way we solve it: We pass the reference from outside and await the future nearly immediately afterwards. As such, the lifetime bounds are upheld and the result we get no longer references `ctx` so we can safely stream it to the consumer.

Due to the limitations in HRTBs, we can't formulate a bound at this time that works with a raw `impl Future<...>` so we need to `Box` it (which creates the monstrosity from your above review comment :sweat_smile:).

Reply:
I am a little confused about this:

> We pass the reference from outside and await the future nearly immediately afterwards. As such, the lifetime bounds are upheld

```rust
    fn user_playlists<'a>(
        &'a self,
        user_id: UserId<'a>,
    ) -> Paginator<'_, ClientResult<SimplifiedPlaylist>> {
        paginate_with_ctx(
            (self, user_id),
            move |(slf, user_id), limit, offset| {
                slf.user_playlists_manual(user_id.as_ref(), Some(limit), Some(offset))
            },
            self.get_config().pagination_chunks,
        )
    }
```

We pass the `(self, user_id)` as the `ctx`, but how does the lifetime bound? 

>  and the result we get no longer references ctx so we can safely stream it to the consumer.

According to the function signature: `for<'ctx> Fn(&'ctx Ctx, u32, u32)`, the closure references to `ctx`, do you mean the result `RequestFuture` doesn't reference to `ctx`?

Reply:
> I am a little confused about this:
> 
> > We pass the reference from outside and await the future nearly immediately afterwards. As such, the lifetime bounds are upheld
> 
> ...
> We pass the `(self, user_id)` as the `ctx`, but how does the lifetime bound?

What I meant by that is:

```rust
pub fn paginate_with_ctx<'a, Ctx: 'a, T, Request>(...) -> Paginator<'a, ClientResult<T>>
where
    T: 'a + Unpin,
    Request: 'a + for<'ctx> Fn(&'ctx Ctx, u32, u32) -> RequestFuture<'ctx, T>,
{
    let mut offset = 0;
    Box::pin(stream! {
        loop {
            // we create a reference to ctx here
            let page = req(&ctx, page_size, offset).await?;
            // due to `await`, the reference is not used again after this line
            offset += page.items.len() as u32;
            for item in page.items {
                // doesn't reference ctx, so yielding the items is fine
                yield Ok(item);
            }
            ...
        }
    })
}
```

> > and the result we get no longer references ctx so we can safely stream it to the consumer.
> 
> According to the function signature: `for<'ctx> Fn(&'ctx Ctx, u32, u32)`, the closure references to `ctx`, do you mean the result `RequestFuture` doesn't reference to `ctx`?

The function itself takes a reference and so does the `RequestFuture`, but when `RequestFuture` is `await`ed, we get a `ClientResult<T>` which does not reference `ctx`.

Reply:
I get your point, it makes sense:

```rust
pub fn paginate_with_ctx<'a, Ctx: 'a, T, Request>(...) -> Paginator<'a, ClientResult<T>>
where
    T: 'a + Unpin,
    Request: 'a + for<'ctx> Fn(&'ctx Ctx, u32, u32) -> RequestFuture<'ctx, T>,
{
    let mut offset = 0;
    Box::pin(stream! {
        loop {
            // we create a reference to ctx here. 
            // the `&ctx` is a shortlived reference, it's just like we put `ctx` into a bounded scope.
           //  ---- scope begins here ----
            let page = req(&ctx, page_size, offset).await?;
            // due to `await`, the reference is not used again after this line
           // ------ scope ends here -----
            offset += page.items.len() as u32;
            for item in page.items {
                // doesn't reference ctx, so yielding the items is fine
                yield Ok(item);
            }
            ...
        }
    })
}
```
 It's clear now :)